src/opm/parser/eclipse/EclipseState/Schedule/WellConnections.cpp

```diff
@@ -100,49 +187,88 @@ namespace Opm {
-            if (satTableIdItem.hasValue(0) && satTableIdItem.get < int > (0) > 0)
-            {
-                satTableId = satTableIdItem.get< int >(0);
-                defaultSatTable = false;
-            }
+        if (satTableIdItem.hasValue(0) && satTableIdItem.get < int > (0) > 0)
+        {
+            satTableId = satTableIdItem.get< int >(0);
+            defaultSatTable = false;
         }
 
-        const WellCompletion::DirectionEnum direction = WellCompletion::DirectionEnumFromString(record.getItem("DIR").getTrimmedString(0));
+        if (diameterItem.hasValue(0))
+            rw = 0.50 * diameterItem.getSIDouble(0);
+        else
+            // The Eclipse100 manual does not specify a default value for the wellbore
+            // diameter, but the Opm codebase has traditionally implemented a default
+            // value of one foot. The same default value is used by Eclipse300.
+            rw = 0.5*unit::feet;
 
         for (int k = K1; k <= K2; k++) {
+            double CF = -1;
+            double Kh = -1;
+
             if (defaultSatTable)
                 satTableId = satnum.iget(grid.getGlobalIndex(I,J,k));
 
             auto same_ijk = [&]( const Connection& c ) {
                 return c.sameCoordinate( I,J,k );
             };
 
+            if (KhItem.hasValue(0) && KhItem.getSIDouble(0) > 0.0)
+                Kh = KhItem.getSIDouble(0);
+
+            if (CFItem.hasValue(0) && CFItem.getSIDouble(0) > 0.0)
+                CF = CFItem.getSIDouble(0);
+
+            /* We start with the absolute happy path; both CF and Kh are explicitly given in the deck. */
+            if (CF > 0 && Kh > 0)
+                goto CF_done;
+
+
+            /* We must calculate CF and Kh from the items in the COMPDAT record and cell properties. */
+            {
+                // Angle of completion exposed to flow.  We assume centre
+                // placement so there's complete exposure (= 2\pi).
+                const double angle = 6.2831853071795864769252867665590057683943387987502116419498;
+                size_t global_index = grid.getGlobalIndex(I,J,k);
+                std::array<double,3> cell_perm = {{ permx[global_index], permy[global_index], permz[global_index]}};
+                std::array<double,3> cell_size = grid.getCellDims(global_index);
+                double r0;
+                const auto& K = permComponents(direction, cell_perm);
+                const auto& D = effectiveExtent(direction, ntg[global_index], cell_size);
+
+                if (r0Item.hasValue(0))
+                    r0 = r0Item.getSIDouble(0);
+                else
+                    r0 = effectiveRadius(K,D);
+
+                if (CF < 0) {
+                    if (Kh < 0)
+                        Kh = std::sqrt(K[0] * K[1]) * D[2];
+                    CF = angle * Kh / (std::log(r0 / std::min(rw, r0)) + skin_factor);
+                } else {
+                    if (KhItem.defaultApplied(0) || KhItem.getSIDouble(0) < 0) {
```

Comment:
Piconit: Why do you check the sign of `.getSIDouble(0)`?  Won't the raw value (`.get<double>(0)`) have the same sign?

Reply:
> Won't the raw value (.get<double>(0)) have the same sign?

Yes it will. Have not really thought about it - but when asked I would say it is a good habit to use the SI accessor; it is only for a few pathological items where the dimension is context dependent the use of raw double access makes sense / is mandated; and these should stand out!?